src/error/mod.rs

```diff
@@ -1,97 +1,167 @@
-        kind: JkErrKind,
-        msg: String,
-        loc: Option<JkErrSpaceLocation>,
-        input: String,
-    ) -> JkError {
-        JkError {
-            kind,
-            msg,
-            loc,
-            input,
+    pub fn with_loc(self, loc: ErrSpaceLocation) -> Error {
+        Error {
+            loc: Some(loc),
+            ..self
         }
     }
 
-    /// Display the error on stderr before exiting the program
     pub fn exit(&self) {
-        eprintln!("{}", self.to_string());
-
         // The exit code depends on the kind of error
         std::process::exit(self.kind as i32 + 1);
     }
+}
 
-    /// What kind of error the error is
-    #[cfg(test)]
-    pub fn kind(&self) -> JkErrKind {
-        self.kind
+/// I/O errors keep their messages
+impl std::convert::From<std::io::Error> for Error {
+    fn from(e: std::io::Error) -> Error {
+        Error::new(ErrKind::IO).with_msg(e.to_string())
     }
+}
 
-    /// Message contained in the error
-    #[cfg(test)]
-    pub fn msg(&self) -> &str {
-        &self.msg
+/// Nom errors are automatically parsing errors
+impl std::convert::From<nom::Err<(&str, nom::error::ErrorKind)>> for Error {
+    fn from(e: nom::Err<(&str, nom::error::ErrorKind)>) -> Error {
+        Error::new(ErrKind::Parsing).with_msg(e.to_string())
     }
 }
 
-impl std::fmt::Display for JkError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        // FIXME: Add better formatting
-        write!(f, "ErrorKind: {:?}\nInfo: {}", self.kind, self.msg.red())
+/// Likewise, if we need to convert from a nom::Err<jinko::Error> to a jinko::Error
+impl std::convert::From<nom::Err<Error>> for Error {
+    fn from(e: nom::Err<Error>) -> Error {
+        match e {
+            nom::Err::Incomplete(_) => Error::new(ErrKind::Parsing),
+            nom::Err::Error(inner) | nom::Err::Failure(inner) => inner,
+        }
     }
 }
 
-impl std::convert::From<std::io::Error> for JkError {
-    fn from(e: std::io::Error) -> Self {
-        JkError::new(JkErrKind::IO, e.to_string(), None, "".to_owned())
+impl nom::error::ParseError<&str> for Error {
+    fn from_error_kind(input: &str, _: nom::error::ErrorKind) -> Error {
+        Error::new(ErrKind::Parsing).with_msg(String::from(input))
     }
-}
 
-// FIXME: Improve formatting, current output is barren
-impl std::convert::From<nom::Err<(&str, nom::error::ErrorKind)>> for JkError {
-    fn from(e: nom::Err<(&str, nom::error::ErrorKind)>) -> Self {
-        JkError::new(JkErrKind::Parsing, e.to_string(), None, "".to_owned())
+    fn append(input: &str, _: nom::error::ErrorKind, _other: Error) -> Error {
+        // FIXME: Should we accumulate errors this way?
+        // let other_msg = match other.msg {
+        //     Some(msg) => format!("{}\n", msg),
+        //     None => String::new(),
+        // };
+
+        Error::new(ErrKind::Parsing).with_msg(String::from(input)) // /* FIXME  */ with_msg(format!("{}{}", other_msg, input))
```

Comment:
When creating new errors, `nom` might try to append to the current error instead of creating a new one. While this sounds great in practice, in jinko right now this means that the same nom error will get displayed multiple times for no apparent reason. Once this is figured out, we should work on accumulating errors properly, the `nom` way